tc/autotuner/options_cache-inl.h

```diff
@@ -0,0 +1,316 @@
+               << filename;
+  } else {
+    toProtobuf().SerializePartialToOstream(&serialized);
+  }
+}
+
+template <typename Backend>
+void OptionsCache<Backend>::recordRuntime(
+    const lang::CanonicalTcString& tc,
+    const std::vector<TensorInfo>& inputs,
+    const std::vector<TensorInfo>& outputs,
+    const std::string& deviceStr,
+    const typename Backend::MappingOptionsType& options,
+    Duration duration) {
+  std::lock_guard<std::mutex> lock(mutex);
+  ++numberCacheAttempts;
+  OptionsCacheKey key{tc, inputs, outputs, deviceStr};
+  bool inserted = false;
+  auto range = store_.equal_range(key);
+  for (auto it = range.first; it != range.second; ++it) {
+    if (it->second.mappingOptions == options) {
+      it->second.runtimes.push_back(duration);
+      inserted = true;
+      break;
+    }
+  }
+  if (!inserted) {
+    store_.emplace(
+        key,
+        OptionsCacheValue<Backend>{std::vector<Duration>{duration}, options});
+  }
+}
+
+template <typename Backend>
+std::vector<typename Backend::MappingOptionsType>
+OptionsCache<Backend>::getTopKOptions(
+    const lang::CanonicalTcString& tc,
+    const std::vector<TensorInfo>& inputs,
+    const std::vector<TensorInfo>& outputs,
+    const std::string& deviceStr,
+    size_t K) const {
+  std::lock_guard<std::mutex> lock(mutex);
+  ++numberAttemptedRetrievals;
+  struct WithTime {
+    typename Backend::MappingOptionsType mappingOptions;
+    Duration duration;
+  };
+  std::vector<WithTime> toSort;
+  OptionsCacheKey key{tc, inputs, outputs, deviceStr};
+  if (store_.count(key) == 0) {
+    return {};
+  }
+  auto range = store_.equal_range(key);
+  for (auto it = range.first; it != range.second; ++it) {
+    if (it->second.runtimes.size() == 0) {
+      throw std::runtime_error("No runtime for cache option");
+    }
+    toSort.push_back(
+        WithTime{it->second.mappingOptions, median(it->second.runtimes)});
+  }
+  // this is of course wasteful but real topk is premature optimization atm
+  std::sort(
+      toSort.begin(), toSort.end(), [](const WithTime& a, const WithTime& b) {
+        return a.duration < b.duration;
+      });
+  std::vector<typename Backend::MappingOptionsType> res;
+  res.reserve(K);
+  for (size_t i = 0; i < std::min(K, toSort.size()); ++i) {
+    res.push_back(toSort[i].mappingOptions);
+  }
+  ++numberSuccessfulRetrievals;
+  return res;
+}
+
+template <typename Backend>
+std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash>
+OptionsCache<Backend>::getKeys() const {
+  std::lock_guard<std::mutex> lock(mutex);
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> keys;
+  for (auto kvp : store_) {
+    if (keys.count(kvp.first) > 0) {
+      continue;
+    }
+    keys.emplace(kvp.first);
+  }
+  return keys;
+}
+
+template <typename Backend>
+void OptionsCache<Backend>::pruneKeepTopK(size_t K) {
+  auto keys = getKeys();
+  std::lock_guard<std::mutex> lock(mutex);
```

Comment:
I find it useful to create a block around a lock and thus highlight the critical section (when it does not span the existing block)